### PR TITLE
Add rollback to qaneeded model

### PR DESCRIPTION
**Pull request by ayush1999**
Fixes: #21

### CODE REVIEW

File: bugbug/models/qaneeded.py [hunk 1]
```diff
@@ -56,6 +56,9 @@ def __init__(self, lemmatization=False):
         self.clf = xgboost.XGBClassifier(n_jobs=16)
         self.clf.set_params(predictor='cpu_predictor')
 
+    def rollback(self, change):
+        return change['field_name'] in ['qawanted', 'qe-verify']
```
Comment (marco-c) [inline bugbug/models/qaneeded.py]: Analyze a few bugs manually, IIRC qe-verify is never in 'field_name'.

Comment (ayush1999) [reply bugbug/models/qaneeded.py]: Yes, neither is `qawanted`. `field_name` contains `qa-contact`, which is the closest it comes.

Comment (marco-c) [reply bugbug/models/qaneeded.py]: Yep. So you'll need to use some other value from `change`.

Comment (ayush1999) [reply bugbug/models/qaneeded.py]: We need to rollback to when `qaneeded` or `qe-verify` was added, right? So maybe setting change['added'] in `['qawanted', 'qe-verify']` should work?

Comment (marco-c) [reply bugbug/models/qaneeded.py]: Have you checked if they can be in change['added']?

Comment (ayush1999) [reply bugbug/models/qaneeded.py]: Yes, they are present in `change['added']`.

Comment (marco-c) [reply bugbug/models/qaneeded.py]: Good, then it should work!
Please also add a test with a bug with qawanted and one with qe-verify.

Comment (marco-c) [reply bugbug/models/qaneeded.py]: Also, if they are present in change['added'], then we should update get_labels to take that into account.

Comment (ayush1999) [reply bugbug/models/qaneeded.py]: `get_labels` already uses this information to give bugs a target label.

**Comment by ayush1999** [comment]
Adding the rollback function as `change['added'] in ['qawanted', 'qe-verify']` actually reduces the accuracy to ~0.73. Is adding this a good option? @marco-c

**Comment by ayush1999** [comment]
Done. Results: The accuracy increases from 0.81 to 0.93.

File: bugbug/models/qaneeded.py [hunk 2]
```diff
@@ -64,7 +67,7 @@ def get_labels(self):
 
             for entry in bug_data['history']:
                 for change in entry['changes']:
-                    if change['added'].startswith('qawanted'):
+                    if change['added'].startswith('qawanted') or change['added'].startswith('qe-verify'):
                         classes[bug_id] = 1
                     elif 'flags' in entry:
```
Comment (marco-c) [inline bugbug/models/qaneeded.py]: The second part here might have the be removed if it is unnecessary. Can you check if there is any bug where this happens?

File: tests/test_qaneeded.py [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+# You can obtain one at http://mozilla.org/MPL/2.0/.
+
+from bugbug.models.qaneeded import QANeededModel
+
+
+def test_get_qaneeded_labels():
+    model = QANeededModel()
+    classes = model.get_labels()
+    assert not classes[1389220]
+    assert classes[1389223]
```
Comment (marco-c) [inline tests/test_qaneeded.py]: Can you also add a test with qawanted?

File: bugbug/models/qaneeded.py (hunk 2 above)
Comment (ayush1999) [reply bugbug/models/qaneeded.py]: Yes, there are quite a few bugs with added as qe-verify.

File: bugbug/models/qaneeded.py (hunk 1 above)
Comment (marco-c) [reply bugbug/models/qaneeded.py]: Looking at the bug you used for the test, this doesn't work. The bug has added == qe-verify+.

File: bugbug/models/qaneeded.py (hunk 2 above)
Comment (marco-c) [reply bugbug/models/qaneeded.py]: I'm not talking about the part you added (we talked about that already, I suggested you to add it :P).
I'm talking about the second part of the pre-existing code, which is analyzing flags.

**Comment by marco-c** [comment]
> Done. Results: The accuracy increases from 0.81 to 0.93.

Mmmh, unexpected! The accuracy in theory should decrease, as we are reducing the information the model can work on.

File: bugbug/models/qaneeded.py (hunk 2 above)
Comment (ayush1999) [reply bugbug/models/qaneeded.py]: Oh! I just checked: there are around 1235 bugs with `name` as `qe-verify`. I think we should let this be here. Right?

File: bugbug/models/qaneeded.py (hunk 1 above)
Comment (ayush1999) [reply bugbug/models/qaneeded.py]: Fixed that. Added the `startswith` method for this.

File: tests/test_qaneeded.py (hunk 3 above)
Comment (ayush1999) [reply tests/test_qaneeded.py]: Done.

File: bugbug/models/qaneeded.py (hunk 2 above)
Comment (marco-c) [reply bugbug/models/qaneeded.py]: Can you show me an example? I want to check the difference between one of them and 1389223.

File: tests/test_qaneeded.py (hunk 3 above)
Comment (marco-c) [reply tests/test_qaneeded.py]: And we should test the case where qe-verify is in the flags.

Also, please add a rollback test.

File: bugbug/models/qaneeded.py (hunk 2 above)
Comment (marco-c) [reply bugbug/models/qaneeded.py]: (Also, if there are cases like that, then you need to add support for it to rollback)

Comment (ayush1999) [reply bugbug/models/qaneeded.py]: > Can you show me an example? I want to check the difference between one of them and 1389223.

Sure, for example bug id 1389252.

Comment (marco-c) [reply bugbug/models/qaneeded.py]: 1389252 doesn't match that: https://bugzilla.mozilla.org/rest/bug/1389252/history.

Comment (ayush1999) [reply bugbug/models/qaneeded.py]: Oh, sorry. I was looking at the wrong `flags` key.

**Comment by ayush1999** [comment]
@marco-c Sorry, I was looking at the wrong `flags` key. I just checked ; the second condition is redundant, I'll remove it from the loop.

**Comment by ayush1999** [comment]
I'e added rollback tests for two bugs (one True and other False). The tracking model also doesn't contain a rollback test. Should I add a test there too? (in this PR?)

**Comment by marco-c** [comment]
> Should I add a test there too? (in this PR?)

No, let's keep this PR about the qaneeded model.

File: bugbug/models/qaneeded.py [hunk 4]
```diff
@@ -56,6 +56,9 @@ def __init__(self, lemmatization=False):
         self.clf = xgboost.XGBClassifier(n_jobs=16)
         self.clf.set_params(predictor='cpu_predictor')
 
+    def rollback(self, change):
+        return change['added'].startswith('qawanted') or change['added'].startswith('qe-verify')
```
Comment (marco-c) [inline bugbug/models/qaneeded.py]: It'd be more Pythonic to do something like:
```Python
any(change['added'].startswith(prefix) for prefix in ['qawanted', 'qe-verify']
```

File: tests/test_qaneeded.py [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+from bugbug import bugzilla
+from bugbug.models.qaneeded import QANeededModel
+
+
+def test_get_qaneeded_labels():
+    model = QANeededModel()
+    classes = model.get_labels()
+    assert not classes[1389220]
+    assert classes[1389223]
+    assert classes[1390433]
```
Comment (marco-c) [inline tests/test_qaneeded.py]: Add a message to the asserts (e.g. Bug should contain qawanted for the qawanted one)

File: tests/test_qaneeded.py [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+
+
+def test_rollback():
+    model = QANeededModel()
+
+    bugs = {}
```
Comment (marco-c) [inline tests/test_qaneeded.py]: Call this `histories`, since you're only storing the history stuff.

File: tests/test_qaneeded.py [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+    for bug in bugzilla.get_bugs():
+        bugs[int(bug['id'])] = bug['history']
+
+    def test(bug_id):
```
Comment (marco-c) [inline tests/test_qaneeded.py]: The name of the function is too generic.

BTW, we also want to test that we are rolling back to the right point in the history

Comment (ayush1999) [reply tests/test_qaneeded.py]: To test if rollback is to the right time, should I check if all comments are made before the `rollback_date`?

Comment (marco-c) [reply tests/test_qaneeded.py]: You could simply assert on the number of changes you loop through before finding the right one.

Comment (marco-c) [reply tests/test_qaneeded.py]: Basically make count a integer increasing it for every iteration of the inner loop, then in the caller do:
`assert rollback_point(BUG_ID) == X`

Comment (ayush1999) [reply tests/test_qaneeded.py]: Sorry for the delay, just finished with my exams. Correct me if I'm wrong, but rollback is supposed to take a bug and return the same bug in an initial stage, right? Because whenever I test the function, the value before and after rollback always comes as equal.

Comment (marco-c) [reply tests/test_qaneeded.py]: Rollback is used to go back in the history of a bug, not exclusively to the initial state, but also to a given state (this is what the `rollback_when` function is for, to know at which point in the history to stop).

In your case, if a bug has qawanted in the history, rollback should go back to the point when qawanted was added.

**Comment by ayush1999** [comment]
I've added a count that increments through all changes and returns whenever `rollback(change)`  is True. Please review @marco-c

File: tests/test_qaneeded.py [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+
+
+def test_rollback():
+    model = QANeededModel()
+
+    histories = {}
+    for bug in bugzilla.get_bugs():
+        histories[int(bug['id'])] = bug
+
+    def contains_field(bug_id):
```
Comment (marco-c) [inline tests/test_qaneeded.py]: This is redundant with the other function, you can remove it and just always use the other function.

File: tests/test_qaneeded.py [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+            for change in history['changes']:
+                if not model.rollback(change):
+                    count += 1
+                else:
+                    ret = True
+                    break
+        return count
+
+    assert contains_field(1390433), 'A bug field should start with qawanted or qe-verify'
+    assert not contains_field(1389136), 'A bug field should start with qawanted or qe-verify'
```
Comment (marco-c) [inline tests/test_qaneeded.py]: Make these two use rollback_point too

File: tests/test_qaneeded.py [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+    assert classes[1389223], 'Bug should contain qawanted in a field'
+    assert classes[1390433], 'Bug should contain qe-verify in a field'
+
+
+def test_rollback():
+    model = QANeededModel()
+
+    histories = {}
+    for bug in bugzilla.get_bugs():
+        histories[int(bug['id'])] = bug
```
Comment (marco-c) [inline tests/test_qaneeded.py]: You were storing the history before, so I suggested to change the name to `histories`. Why did you move to storing the whole bug instead?

Comment (ayush1999) [reply tests/test_qaneeded.py]: Removed it. I thought it'd maintain uniformity, but it seems redundant now.

File: tests/test_qaneeded.py [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+        histories[int(bug['id'])] = bug['history']
+
+    def rollback_point(bug_id):
+        count = 0
+        ret = False
+        for history in histories[bug_id]:
+            if ret:
+                break
+            for change in history['changes']:
+                if not model.rollback(change):
```
Comment (marco-c) [inline tests/test_qaneeded.py]: This can be simplified:
```Python
if model.rollback(change):
    return count
count += 1
```
And you can completely remove the `ret` variable.

**Comment by ayush1999** [comment]
Done. Please review.
